### PR TITLE
error reporting with error mappers

### PR DESCRIPTION
**Pull request by yurii-klymenko**
reworked error reporting, 
- errors codes < 1000 belong stix_transmission, they are general, about connectivity and certificates, configuration etc. Potentially they will be mapped to http 5xx
- error codes >= 1000 belong to modules. Potentially they will be mapped to http 4xx.

### CODE REVIEW

File: stix_shifter/utils/error_response.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+class ErrorCode(Enum):
+    TRANSMISSION_UNKNOWN = 500
+    TRANSMISSION_CONNECT = 600
+    TRANSMISSION_AUTH_SSL = 610
+    TRANSMISSION_AUTH_CREDENTIALS = 620
+    TRANSMISSION_MODULE_DEFAULT_ERROR = 700
+    TRANSMISSION_QUERY_PARSING_ERROR = 802
+    TRANSMISSION_QUERY_LOGICAL_ERROR = 803
+    TRANSMISSION_RESPONSE_EMPTY_RESULT = 903
```
Comment (jleemur) [inline stix_shifter/utils/error_response.py]: How did you decide on error codes? i.e. starting at 802, 903, etc.

Comment (yurii-klymenko) [reply stix_shifter/utils/error_response.py]: 5xx - transmission unexpected error
6xx - connectivity errors
7xx - submodule unhandled errors
8xx - "error in query" error
9xx - special case for remote system errors
10xx - non fatal-, retry-later-  errors

if you do not like them sometimes starting from xx[^0] - just let me know.

File: stix_shifter/stix_transmission/src/modules/bigfix/bigfix_results_connector.py [hunk 2]
```diff
@@ -1,27 +1,44 @@
...
+import xmltodict
 
+class UnexpectedResponseException(Exception):
+    pass
 
 class BigFixResultsConnector(BaseResultsConnector):
     def __init__(self, api_client):
         self.api_client = api_client
 
+    def get_succes_status(self, data_dict):
```
Comment (jleemur) [inline stix_shifter/stix_transmission/src/modules/bigfix/bigfix_results_connector.py]: success*

File: stix_shifter/utils/error_response.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            if (isinstance(message_struct, collections.Mapping) or type(message_struct).__name__=='list'):
+                struct = message_struct.copy()
+                for i in message_path:
+                    if (isinstance(struct, collections.Mapping) and i in struct) or (type(struct).__name__=='list' and isinstance(i, int) and i < len(struct)):
+                        struct = struct[i]
+                        if struct is None:
+                            break
+                    elif i[:1] == '+' and isinstance(struct, list):
+                        key, value = i[1:].split('=')
+                        filted_struct = list(filter(lambda item: (key in item and str(item[key])==str(value)), struct))
```
Comment (jleemur) [inline stix_shifter/utils/error_response.py]: filtered*

File: stix_shifter/utils/error_response.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                error_msg = str(error)
+
+            if len(error_msg) > 0:
+                if len(message) > 0:
+                    message += '; '
+                message += error_msg
+
+        if message is not None and len(message)>0:
+            if error_code.value == ErrorCode.TRANSMISSION_UNKNOWN.value:
+                if 'uthenticat' in message or 'uthoriz' in message:
```
Comment (jleemur) [inline stix_shifter/utils/error_response.py]: authenticate* authorize* ?

File: stix_shifter/stix_transmission/src/modules/splunk/splunk_error_mapper.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+from .....utils.error_mapper_base import ErrorMapperBase
+from .....utils.error_response import ErrorCode
+
+error_mapping = {
+        'Unknown sid.': ErrorCode.TRANSMISSION_SEARCH_DOES_NOT_EXISTS,
```
Comment (jleemur) [inline stix_shifter/stix_transmission/src/modules/splunk/splunk_error_mapper.py]: Could you make these a constant defined in ErrorCode?
`Unknown sid.` and `Unable to parse the search`

File: stix_shifter/stix_transmission/src/modules/splunk/splunk_results_connector.py [hunk 6]
```diff
@@ -9,26 +10,19 @@ def __init__(self, api_client):
...
-                return_obj['data'] = results
-            else:
-                return_obj['success'] = False
-                return_obj['error'] = response_json['messages'][0]['text']
-            return return_obj
+        response = self.api_client.get_search_results(search_id, offset, length)
+        response_code = response.code
+        response_json = json.load(response)
+        if "results" in response_json:
+            results = [] if (response_json['results'] == []) else response_json['results']
```
Comment (jleemur) [inline stix_shifter/stix_transmission/src/modules/splunk/splunk_results_connector.py]: No need for ternary, `results = response_json['results']`

File: stix_shifter/utils/error_response.py (hunk 4 above)
Comment (yurii-klymenko) [reply stix_shifter/utils/error_response.py]: upper case and lower case for the first letter with one pattern

File: stix_shifter/stix_transmission/src/modules/splunk/splunk_error_mapper.py (hunk 5 above)
Comment (yurii-klymenko) [reply stix_shifter/stix_transmission/src/modules/splunk/splunk_error_mapper.py]: It is a module specific stuff. These lines are used in the splunk  module only.

File: stix_shifter/utils/error_response.py (hunk 1 above)
Comment (jleemur) [reply stix_shifter/utils/error_response.py]: I'm still confused with error codes? Why don't you..

```
    TRANSMISSION_UNKNOWN = 500
    TRANSMISSION_CONNECT = 600
    TRANSMISSION_AUTH_SSL = 601
    TRANSMISSION_AUTH_CREDENTIALS = 602
    TRANSMISSION_MODULE_DEFAULT_ERROR = 700
    TRANSMISSION_QUERY_PARSING_ERROR = 800
    TRANSMISSION_QUERY_LOGICAL_ERROR = 801
    TRANSMISSION_RESPONSE_EMPTY_RESULT = 900
    TRANSMISSION_SEARCH_DOES_NOT_EXISTS = 901
    TRANSMISSION_INVALID_PARAMETER = 902
    TRANSMISSION_REMOTE_SYSTEM_IS_UNAVAILABLE = 1000
```

**Comment by jleemur** [comment]
LGTM 🥇

**Pull request merged**